### ISSUE TITLE
Conda API in a long lived process

### ISSUE THREAD

**Issue by shughes-uk** [conda/conda#10762]
Hiya from the Coiled team! 

## Current Behavior

Using `Conda.api.Solver` in a long lived process, it seems to cache package availability in memory. Recreating this is very tricky but it pops up for us when the following happens. This only occurs if it happens in the lifetime of one process:
* Attempt to solve a conda environment with a version of a package that is not available on conda-forge (most often because it has been released but does not seem to have made it through the CDN yet)
* Version becomes available
* Attempts to solve the previous conda environment still fail
* Restarting the process causes the environment to start resolving correctly

This has been a head scratcher for us but we suspect there is some sort of in-memory caching going on that is causing it. As a workaround we're using `multiprocessing` to have all our solves done in short lived processes, but it would be nice to either have the ability to bust whatever is caching is happening, or to turn it off entirely.
## Environment Information
<details open><summary><code>`conda info`</code></summary><p>

```
     active environment : base
    active env location : /opt/conda
            shell level : 1
       user config file : /root/.condarc
 populated config files :
          conda version : 4.10.1
    conda-build version : not installed
         python version : 3.8.8.final.0
       virtual packages : __linux=4.14.232=0
                          __glibc=2.28=0
                          __unix=0=0
                          __archspec=1=x86_64
       base environment : /opt/conda  (writable)
      conda av data dir : /opt/conda/etc/conda
  conda av metadata url : https://repo.anaconda.com/pkgs/main
           channel URLs : https://repo.anaconda.com/pkgs/main/linux-64
                          https://repo.anaconda.com/pkgs/main/noarch
                          https://repo.anaconda.com/pkgs/r/linux-64
                          https://repo.anaconda.com/pkgs/r/noarch
          package cache : /opt/conda/pkgs
                          /root/.conda/pkgs
       envs directories : /opt/conda/envs
                          /root/.conda/envs
               platform : linux-64
             user-agent : conda/4.10.1 requests/2.25.1 CPython/3.8.8 Linux/4.14.232-177.418.amzn2.x86_64 debian/10 glibc/2.28
                UID:GID : 0:0
             netrc file : None
           offline mode : False
```
</p></details>

**Comment by necaris** [comment]
@kalefranz @msarahan any chance this is something you can shed light on?

**Comment by jezdez** [comment]
@shughes-uk There are a number of caching techniques used in Conda, e.g. `@memoize`/`@memoizemethod`, the `conda.api.PackageCacheData`, `conda.api.SubdirData` etc.

A few ideas that you could try:

- Call `api.PackageCacheData("<pkg_dirs_path>").reload()` to reset the in-memory cache for package data.

- Call `api.SubDirData("<channel_URL>").reload()` to reset the internal cache of subdir data.

You can get the parameters for both using `conda info --json`.

**Comment by ian-r-rose** [comment]
I was able to reproduce this locally today:

1. Go to [`conda-forge`](https://anaconda.org/conda-forge/repo), select one of the recently published packages (typically one within the hour). I don't think this one needs to be recent, we just need to make sure the repodata is read at least once, but it's what I did (I chose `mongodb=5.0.0`)
2. Open an (i)python session, and enter the following, which should work:
```python
import conda.api
solver = conda.api.Solver(prefix="not-an-env", channels=["conda-forge"], specs_to_add=["mongodb=5.0.0"])
solver.solve_for_final_state()
```
3. Do not close your python session. Go make a cup of coffee, wait for another package to be published to `conda-forge`
4. Go make lunch. That package needs to show up in the platform subdir for the channel, e.g. [noarch](https://conda.anaconda.org/conda-forge/noarch/), which can take another while.
4. Once a new package is available in the listing (I got `dotnet=5.0.301`), create a new solver and try to solve. It fails.
```python
solver = conda.api.Solver(prefix="not-an-env", channels=["conda-forge"], specs_to_add=["dotnet=5.0.301"])
solver.solve_for_final_state()
```
```python-traceback
...
ResolvePackageNotFound: 
  - dotnet=5.0.301
```
5. Try out @jezdez 's suggestion above. It succeeds.
```python
conda.api.SubdirData(conda.api.Channel("conda-forge/linux-64")).reload()
solver = conda.api.Solver(prefix="not-an-env", channels=["conda-forge"], specs_to_add=["dotnet=5.0.301"])
solver.solve_for_final_state()
```

So, seems like a fairly straightforward instance of the cache needing to be reloaded, though in our case it was tough to track down since it was in such a long running process and the cache could get quite stale.
